.pre-commit-config.yaml

```diff
@@ -0,0 +1,114 @@
+        language: system
+        pass_filenames: false
+
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.0.1
+    hooks:
+      - id: trailing-whitespace
+        args: [--markdown-linebreak-ext=md]
+      - id: check-executables-have-shebangs
+      - id: check-json
+        exclude: tsconfig.json # contains comments
+      - id: check-case-conflict
+      - id: check-toml
+      - id: check-merge-conflict
+      - id: check-xml
+      - id: check-yaml
+      - id: end-of-file-fixer
+        exclude: test/tapes/.+\.json5
+      - id: check-symlinks
+      - id: mixed-line-ending
+      - id: fix-encoding-pragma
+        args:
+          - --remove
+      - id: requirements-txt-fixer
+
+  - repo: https://github.com/PyCQA/isort
+    rev: 5.9.1
+    hooks:
+      - id: isort
+        files: \.py$
+        exclude: ^build/.*$|^.tox/.*$|^venv/.*$
+        args:
+          - --lines-after-imports=2
+          - --multi-line=3
+          - --trailing-comma
+          - --force-grid-wrap=0
+          - --use-parentheses
+          - --ensure-newline-before-comments
+          - --line-length=88
+
+  - repo: https://github.com/PyCQA/flake8
+    rev: 3.9.2
+    hooks:
+      - id: flake8
+        args:
+          - --config=automations/.flake8
+
+  - repo: https://github.com/ambv/black
+    rev: 22.3.0
+    hooks:
+      - id: black
+        args:
+          - --safe
+
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: v2.5.0
+    hooks:
+      - id: prettier
+        additional_dependencies:
+          - prettier@2.2.1
+          - prettier-plugin-tailwindcss@0.1.12
+
+  - repo: https://github.com/pre-commit/mirrors-eslint
+    rev: v8.3.0
+    hooks:
+      - id: eslint
+        files: ^.(js|ts|vue)$
+        args:
+          - --ignore-path=.gitignore
+          - --ignore-path=.eslintignore
+          - --max-warnings=0
+        additional_dependencies:
+          - '@babel/eslint-parser@7.17.0'
+          - '@intlify/eslint-plugin-vue-i18n@1.4.0'
+          - '@typescript-eslint/eslint-plugin@5.20.0'
+          - '@typescript-eslint/parser@5.20.0'
+          - eslint-config-prettier@8.5.0
+          - eslint-import-resolver-custom-alias@1.3.0
+          - eslint-plugin-eslint-comments@3.2.0
+          - eslint-plugin-import@2.26.0
+          - eslint-plugin-prettier@4.0.0
+          - eslint-plugin-tsdoc@0.2.16
+          - eslint-plugin-unicorn@42.0.0
+          - eslint-plugin-vue@9.0.0
+          - eslint-plugin-vuejs-accessibility@1.1.1
+
+  - repo: https://github.com/koalaman/shellcheck-precommit
+    rev: v0.8.0
+    hooks:
+      - id: shellcheck
+
+  - repo: local
+    hooks:
+      - id: types
+        name: types
+        entry: bash -c 'pnpm types'
+        language: system
+        pass_filenames: false
+
+      - id: test:unit
+        name: test:unit
+        entry: bash -c 'pnpm test:unit'
+        language: system
+        pass_filenames: false
+        stages:
+          - push
```

Comment:
TIL about `repo: local`, so cool!

Reply:
It was fascinating to see `pre-commit` having workarounds for so many scenarios I thought would be edge-cases or unsupported.